### ISSUE TITLE
Reader outer_xml is decoding escaped entities in JRuby

### ISSUE THREAD

**Issue by jcronk** [sparklemotion/nokogiri#1523]
Example:

``` ruby
require 'nokogiri'

reader = Nokogiri::XML::Reader(File.open 'entities.xml')
reader.each do |node|
  puts node.outer_xml if node.node_type == 1 && node.name == 'Main'
end
```

using this XML:

``` xml
<Root>
    <Main>
        <Element>This has &amp; in it</Element>
    </Main>
    <Main>
        <Element>This has &gt; in it</Element>
    </Main>
    <Main>
        <Element>This has &lt; in it</Element>
    </Main>
</Root>
```

Produces this output:

``` xml
<Main>
        <Element>This has & in it</Element>
    </Main>
<Main>
        <Element>This has > in it</Element>
    </Main>
<Main>
        <Element>This has < in it</Element>
    </Main>
```

This makes it impossible to pass `outer_xml` to `Nokogiri::XML()` because with strict mode on, it will throw an error, and with strict off, it will drop parts of the XML.  I tested this on jruby-1.7.24 and jruby-9.0.5.0.

**Comment by flavorjones** [comment]
Confirmed, MRI generates this output in the above test case:

```
<Main>
        <Element>This has &amp; in it</Element>
    </Main>
<Main>
        <Element>This has &gt; in it</Element>
    </Main>
<Main>
        <Element>This has &lt; in it</Element>
    </Main>
```

while JRuby generates:

```
<Main>
        <Element>This has & in it</Element>
    </Main>
<Main>
        <Element>This has > in it</Element>
    </Main>
<Main>
        <Element>This has < in it</Element>
    </Main>
```

**Comment by rob99** [comment]
I encountered this problem in 1.7.0.1 but was able to fix it by upgrading to 1.8.2.

**Comment by flavorjones** [comment]
Worth noting: `Reader#inner_xml` has the same issue on JRuby.

```ruby
#! /usr/bin/env ruby

require "bundler/inline"

gemfile do
  source "https://rubygems.org"
  gem "nokogiri", path: "."
end

require 'nokogiri'

xml = <<~XML
<Root>
    <Main>
        <Element>This has &amp; in it</Element>
    </Main>
    <Main>
        <Element>This has &gt; in it</Element>
    </Main>
    <Main>
        <Element>This has &lt; in it</Element>
    </Main>
</Root>
XML

outer_xml = +""
reader = Nokogiri::XML::Reader(xml)
reader.each do |node|
  outer_xml << node.outer_xml if node.node_type == 1 && node.name == 'Main'
end

pp ["outer:", outer_xml]

inner_xml = +""
reader = Nokogiri::XML::Reader(xml)
reader.each do |node|
  inner_xml << node.inner_xml if node.node_type == 1 && node.name == 'Main'
end

pp ["inner:", inner_xml]

unless outer_xml.include?("&amp;") && inner_xml.include?("&amp;")
  puts "Failed"
  exit 1
end

puts "OK"
```

**Comment by flavorjones** [comment]
Proposed fix is in #3268
